src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala

```diff
@@ -120,25 +98,51 @@ object JsonParser {
 
   case class JsonSimpleInstruction(
       prim: String,
-      args: Option[List[Either[JsonExpression, List[JsonInstruction]]]] = None,
+      args: Option[List[EmbeddedElement]] = None,
       annots: Option[List[String]] = None
   ) extends JsonInstruction {
     override def toMichelsonInstruction =
       MichelsonSingleInstruction(
         name = prim,
         annotations = annots.getOrElse(List.empty),
-        embeddedElements = args.getOrElse(List.empty).map {
-          case Left(jsonExpression) => jsonExpression.toMichelsonExpression
-          case Right(Nil) => MichelsonEmptyInstruction
-          case Right(jsonInstructions) => MichelsonInstructionSequence(jsonInstructions.map(_.toMichelsonInstruction))
-        }
+        embeddedElements = args.getOrElse(List.empty).map(toMichelsonElement)
       )
   }
 
   case class JsonInstructionSequence(instructions: List[JsonInstruction]) extends JsonInstruction {
     override def toMichelsonInstruction = MichelsonInstructionSequence(instructions.map(_.toMichelsonInstruction))
   }
 
+  /*
```

Comment:
There are other types, `address`, `nat`, boolean, to name a few, do they need to be implemented separately?

Reply:
I'm not sure how the types are represented because it's not documented so I created those structures only for `int`, `string` and `bytes`. They were the only constructions for which parsing failed.